crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2842,16 +2865,42 @@ enum InvalidTypeExpression {
     ClassVarInTypeExpression,
     /// The `Final` type qualifier was used in a type expression
     FinalInTypeExpression,
+    /// Dynamic variables are not allowed in type expressions
+    DynamicVariable(Type<'db>),
```

Comment:
I'm not sure about this "dynamic variable" naming. I'd rather not introduce that concept/wording without a clear principled definition of what constitutes a "dynamic variable" and how we would distinguish one from a simple implicit type alias.

What this PR actually implements is not any logic at all related to what is a "dynamic variable", just logic about what types are valid in annotations. So we have no idea how "dynamic" the variable is or isn't, just that it was of a type that isn't valid in a type expression.

(For example, an unconditional global `x = 1` is no more or less a "dynamic variable" than `x = int` is, but the former would raise this error if used in an annotation, while the latter is a valid implicit type alias. We aren't making any distinctions here based on "how dynamic", just based on the type.)
```suggestion
    /// Some types are not allowed in type expressions
    InvalidType(Type<'db>),
```